### PR TITLE
Set a finalizer function on the transport to close idle connections.

### PR DESCRIPTION
**Pull request by jefferai**
This ensures that we do not leak fds, potentially running out of them.

### CODE REVIEW

**Comment by jefferai** [comment]
LGTM via https://github.com/fsouza/go-dockerclient/pull/434#issuecomment-165563345

**Comment by orivej** [comment]
The patch looks good and works for me in go-dockerclient. New functions should be documented.

File: cleanhttp.go [hunk 1]
```diff
@@ -26,3 +29,11 @@ func DefaultClient() *http.Client {
 		Transport: DefaultTransport(),
 	}
 }
+
+func SetFinalizer(transport *http.Transport) {
+	runtime.SetFinalizer(&transport, FinalizeTransport)
```
Comment (mitchellh) [inline cleanhttp.go]: This needs docs, and I'd rename this to `SetTransportFinalizer`.

File: cleanhttp.go [hunk 2]
```diff
@@ -26,3 +29,11 @@ func DefaultClient() *http.Client {
...
+}
+
+func FinalizeTransport(t **http.Transport) {
+	(*t).CloseIdleConnections()
```
Comment (mitchellh) [inline cleanhttp.go]: Is there any reason to expose this function? why don't we just inline it above into SetFinalizer. I don't see a reason (esp with the weird function signature) for this to exist on its own as a public API.

Comment (jefferai) [inline cleanhttp.go]: Mostly I didn't see a reason *not* to expose it. I can make it private. I'll test; if there are no wonky behaviors I'll inline it.

Comment (mitchellh) [inline cleanhttp.go]: Let's inline it. As we note "shipping is forever." especially with Go. So let's error on the side of smaller API surface, and only expose as needed, though I doubt it in this case.

**Pull request merged**